src/connector/src/parser/chunk_builder.rs

```diff
@@ -14,83 +14,201 @@
+            // Note this
+            self.finish_current_chunk();
+        }
+
+        _ = self
+            .row_writer()
+            .invisible()
+            .with_meta(meta)
+            .do_insert(|_| Ok(Datum::None));
+        self.finish_current_chunk(); // each heartbeat should be a separate chunk
+    }
+
+    /// Finish and build a [`StreamChunk`] from the current pending records in the builder,
+    /// no matter whether the builder is in a transaction or not, `split_txn` or not. The
+    /// built chunk will be appended to the `ready_chunks` and the builder will be reset.
+    pub fn finish_current_chunk(&mut self) {
+        if self.op_builder.is_empty() {
+            return;
+        }
+
+        let (builders, op_builder, vis_builder) =
+            Self::create_builders(&self.column_descs, self.source_ctrl_opts.chunk_size);
+        let chunk = StreamChunk::with_visibility(
+            std::mem::replace(&mut self.op_builder, op_builder),
+            std::mem::replace(&mut self.builders, builders)
                 .into_iter()
                 .map(|builder| builder.finish().into())
                 .collect(),
-            self.vis_builder.finish(),
-        )
+            std::mem::replace(&mut self.vis_builder, vis_builder).finish(),
+        );
+        self.ready_chunks.push(chunk);
+
+        if let Some(ref mut txn) = self.ongoing_txn {
+            tracing::warn!(
+                txn_id = txn.id,
+                len = txn.len,
+                "splitting an ongoing transaction"
+            );
+            txn.len = 0;
+        }
     }
 
-    /// Resets the builder and returns a [`StreamChunk`], while reserving `next_cap` capacity for
-    /// the builders of the next [`StreamChunk`].
-    #[must_use]
-    pub fn take_and_reserve(&mut self, next_cap: usize) -> StreamChunk {
-        let descs = std::mem::take(&mut self.descs); // we don't use `descs` in `finish`
-        let builder = std::mem::replace(self, Self::with_capacity(descs, next_cap));
-        builder.finish()
+    /// Consumes and returns the ready [`StreamChunk`]s.
+    pub fn consume_ready_chunks(&mut self) -> impl ExactSizeIterator<Item = StreamChunk> + '_ {
+        self.ready_chunks.drain(..)
     }
 
-    pub fn len(&self) -> usize {
+    fn current_chunk_len(&self) -> usize {
         self.op_builder.len()
     }
 
-    pub fn is_empty(&self) -> bool {
-        self.op_builder.is_empty()
+    /// Commit a newly-written record by appending `op` and `vis` to the corresponding builders.
+    /// This is supposed to be called via the `row_writer` only.
+    fn commit_record(&mut self, op: Op, vis: bool) {
+        self.op_builder.push(op);
+        self.vis_builder.append(vis);
+
+        let curr_chunk_size = self.current_chunk_len();
+        let max_chunk_size = self.source_ctrl_opts.chunk_size;
+
+        if let Some(ref mut txn) = self.ongoing_txn {
+            txn.len += 1;
+
+            if txn.len >= MAX_TRANSACTION_SIZE
```

Comment:
There was a warning log before. Shall we keep it?

Reply:
The warning log is moved to `finish_current_chunk`, so that we won't miss any call site of `finish_current_chunk` that can cause an interrupt of transaction.